ops/framework.py

```diff
@@ -1152,6 +1147,16 @@ def __init__(self, parent, key):
             def _on_seen(self, event):
                 self._stored.seen.add(event.uuid)
 
+    Data is stored alongside the charm (in the charm container for Kubernetes
+    sidecar charms, and on the machine for machine charms), except for podspec
+    charms and charms explicitly passing `True` for `use_juju_for_storage` when
+    running :meth:`ops.main` (in those cases, the data is stored in Juju).
+
+    For machine charms, charms are upgraded in-place on the machine, so the data
+    is preserved. For Kubernetes sidecar charms, when the charm is upgraded, the
+    pod is replaced, so any data is lost. When data should be preserved across
+    upgrades, Kubernetes sidecar charms should use a peer-relation for the data
+    instead of `StoredState`.
```

Comment:
Random ideas -- I know that `use_juju_for_storage` isn't actually going away with 4.0, but do we want to continue to steer people away from it?

We could mention podspec charms going away in future? e.g.

"except for podspec  charms (deprecated in favour of sidecare charms), where the data is stored in Juju."

And maybe move the bit about `use_juju_for_storage` to *after* the good recommendation in the last paragraph for preserving data across upgrades? e.g. as a final sentence:

"To explicitly opt into Juju data storage, set `use_juju_for_storage` to `True` when running :meth:`ops.main` (but this is not recommended)."

Reply:
podspec charms are deprecated, aren't they? Second guessing myself ..

Reply:
> podspec charms are deprecated, aren't they? Second guessing myself ..

Yes :). You have to find that info [in the Juju docs, though](https://juju.is/docs/sdk/charm-taxonomy#heading--podspec).

Reply:
I've adjusted it to mention podspec and `use_juju_for_storage` being deprecated - I agree that's worth doing.

I don't think the `use_juju_for_storage` mention should be at the end, though - even with a "this is not recommended" I think it promotes it too much saying that it's a way to opt in. We don't want people doing this (and they'll get a warning if they do) - it's mentioned here so people that are doing it know that it is a consequence, but if they want the storage to be in Juju then they should use a peer relation.

Reply:
Makes sense!